### PR TITLE
Fix set semantics bug

### LINKED ISSUE #6: PendingStream and internal store get out of sync due to set-semantics
An RDF store assumes set-semantics meaning it will ignore duplicate triples. Currently, the PendingStream does not comply to set-semantics and will push duplicate triples.

In essence, a match call to the store before and after a duplicate triple will not be the same.

### PR DESCRIPTION
**Pull request by maartyman**

### CODE REVIEW

**Comment by maartyman** [comment]
High level overview:
1. Create new PassThrough stream: `storeImportStream`.
2. Read the import stream in on-demand mode. For each value check if it has a match.
3. Await match if it matches, do nothing. If it doesn't match, push into `pendingStream` and `storeImportStream`.
4.  Import `storeImportStream` into store.

**Comment by maartyman** [comment]
A clean-up of the implementation still needs to be done!

File: test/StreamingStore-test.ts [hunk 1]
```diff
@@ -465,4 +465,15 @@ describe('StreamingStore', () => {
...
+
+  it('handles duplicates in import (set-semantics)', async() => {
+    const match = store.match();
+    await promisifyEventEmitter(store.import(streamifyArray([
+      quad('s1', 'p1', 'o1'),
+      quad('s1', 'p1', 'o1'),
+    ])));
+    store.end();
+
+    expect(await arrayifyStream(match)).toEqualRdfQuadArray(await arrayifyStream(store.getStore().match()));
```
Comment (rubensworks) [inline test/StreamingStore-test.ts]: Can we consider making the expected triples more explicit?
This makes the test slightly more easily understandable IMO:
```suggestion
    expect(await arrayifyStream(match)).toEqualRdfQuadArray([
      quad('s1', 'p1', 'o1'),
    ]);
```

File: lib/StreamingStore.ts [hunk 2]
```diff
@@ -40,24 +40,68 @@ implements RDF.Source<Q>, RDF.Sink<RDF.Stream<Q>, EventEmitter> {
...
   }
 
-  protected importToListeners(stream: RDF.Stream<Q>): void {
-    stream.on('data', (quad: Q) => {
-      for (const pendingStream of this.pendingStreams.getPendingStreamsForQuad(quad)) {
-        if (!this.ended) {
-          pendingStream.push(quad);
-          pendingStream.emit('quad', quad);
-        }
+  private onReadable(stream: RDF.Stream<Q>, storeImportStream: PassThrough): void {
```
Comment (rubensworks) [inline lib/StreamingStore.ts]: `stream` may contain errors (e.g. when it originates from a parser), so we should look into forwarding those error events to `storeImportStream` I suspect.

File: lib/StreamingStore.ts [hunk 3]
```diff
@@ -40,24 +40,68 @@ implements RDF.Source<Q>, RDF.Sink<RDF.Stream<Q>, EventEmitter> {
...
+          matchStream.once('end', () => {
+            if (!this.ended) {
+              storeImportStream.push(staticQuad);
+              for (const pendingStream of this.pendingStreams.getPendingStreamsForQuad(staticQuad)) {
+                pendingStream.push(staticQuad);
+                pendingStream.emit('quad', staticQuad);
+              }
+            }
+            if (streamEnded) {
+              storeImportStream.end();
```
Comment (rubensworks) [inline lib/StreamingStore.ts]: Wondering if this is the desired behaviour here.

Let's say our internal store is very slow, and `matchStream` ends very late, later than `stream`.
According to the logic above, late arriving quads will not be considered anymore, while I suspect we still want those slow quads to be considered as quads that do not have to be pushed to the pending streams?

Comment (rubensworks) [reply lib/StreamingStore.ts]: (We probably need to test these edge-cases as well with some unit tests 😅 )

File: lib/StreamingStore.ts [hunk 4]
```diff
@@ -40,24 +40,68 @@ implements RDF.Source<Q>, RDF.Sink<RDF.Stream<Q>, EventEmitter> {
...
+        await new Promise<void>(resolve => {
+          const matchStream = this.store.match(
+            staticQuad.subject,
+            staticQuad.predicate,
+            staticQuad.object,
+            staticQuad.graph,
+          );
+
+          matchStream.once('data', () => {
+            matchStream.removeAllListeners();
```
Comment (rubensworks) [inline lib/StreamingStore.ts]: It might be safer to only remove the current listener.
It could be for instance that the store attaches some internal listeners to the returned stream (or the stream is reused somehow), so we probably don't want this to to break.

File: lib/StreamingStore.ts (hunk 2 above)
Comment (maartyman) [reply lib/StreamingStore.ts]: Yes, shouldn't these errors also be propagated to the `pendingStreams`?

Comment (rubensworks) [reply lib/StreamingStore.ts]: I think they should not, because the error will be forwarded to `innerstore.import`, which in its turn returns an `EventEmitter` that should forward the error as well. So the error will end up at the caller of `streamingstore.import`.

**Comment by maartyman** [comment]
Everything above has been implemented. Are more tests needed?

File: lib/StreamingStore.ts [hunk 5]
```diff
@@ -40,14 +40,50 @@ implements RDF.Source<Q>, RDF.Sink<RDF.Stream<Q>, EventEmitter> {
...
   }
 
-  protected importToListeners(stream: RDF.Stream<Q>): void {
-    stream.on('data', (quad: Q) => {
-      for (const pendingStream of this.pendingStreams.getPendingStreamsForQuad(quad)) {
-        if (!this.ended) {
-          pendingStream.push(quad);
-          pendingStream.emit('quad', quad);
-        }
+  protected importToListeners(stream: RDF.Stream<Q>, storeImportStream: PassThrough): void {
```
Comment (rubensworks) [inline lib/StreamingStore.ts:49]: Could you add a jsdoc comment on this method to briefly explain what is happening in this method?
It used to be a trivial method, but it's definitely not the case anymore :-)

Perhaps some inline comments above non-trivial code blocks can also help, similar to what I did here: https://github.com/comunica/comunica/blob/master/packages/actor-optimize-query-operation-join-connected/lib/ActorOptimizeQueryOperationJoinConnected.ts#L35-L52

**Review by rubensworks — CHANGES_REQUESTED** [review]
Besides the docs comments, everything looks fine, great tests as well 👍

**Comment by rubensworks** [comment]
@surilindur Before we merge this, is it easy for you to test this in your setup with the aggregate store enabled? If so, can you check if this change solves your problem? If not, you can ignore this comment, and we can check it later for you.

**Comment by surilindur** [comment]
Thank you for looking into a fix!

After changing to the branch here, I stopped getting results from any queries with the aggregate store enabled. After some additional investigation, I found out that using the branch here in base Comunica (master branch) causes this set of unit tests to fail: `packages/actor-rdf-resolve-quad-pattern-hypermedia/test/MediatedQuadSource-test.ts`

The failure seems to indicate missing quads from a match call (the same test file has two other failing tests):

```
Summary of all failing tests
 FAIL  packages/actor-rdf-resolve-quad-pattern-hypermedia/test/MediatedQuadSource-test.ts (11.457 s)
  ● MediatedQuadSource › A MediatedQuadSource instance with aggregated store › match › should match three chained sources when queried multiple times

    expected two graphs to be isomorphic.

      Expected:
    [
      {"subject":"s11","predicate":"p11","object":"o11","graph":""},
      {"subject":"s21","predicate":"p21","object":"o21","graph":""},
      {"subject":"s12","predicate":"p12","object":"o12","graph":""},
      {"subject":"s22","predicate":"p22","object":"o22","graph":""},
      {"subject":"s13","predicate":"p13","object":"o13","graph":""},
      {"subject":"s23","predicate":"p23","object":"o23","graph":""}
    ]

      Actual:
    [

    ]

    Missing Quads (that don't contain Blank Nodes):
    [
      {"subject":"s11","predicate":"p11","object":"o11","graph":""},
      {"subject":"s21","predicate":"p21","object":"o21","graph":""},
      {"subject":"s12","predicate":"p12","object":"o12","graph":""},
      {"subject":"s22","predicate":"p22","object":"o22","graph":""},
      {"subject":"s13","predicate":"p13","object":"o13","graph":""},
      {"subject":"s23","predicate":"p23","object":"o23","graph":""}
    ]

    Additional Quads (that don't contain Blank Nodes):
    [

    ]

    Missing Blank Node Patterns:


    Additional Blank Node Patterns:

      254 |         ]);
      255 |         expect(mediatorRdfResolveHypermedia.mediate).toHaveBeenCalledTimes(4);
    > 256 |         expect(await arrayifyStream(source.match(v, v, undefined!, v, context))).toBeRdfIsomorphic([
          |                                                                                  ^
      257 |           quad('s11', 'p11', 'o11'),
      258 |           quad('s21', 'p21', 'o21'),
      259 |           quad('s12', 'p12', 'o12'),

      at Object.<anonymous> (packages/actor-rdf-resolve-quad-pattern-hypermedia/test/MediatedQuadSource-test.ts:256:82)
```
If I find the time, I can look into what is going wrong, but if someone else has the time before it, that would be amazing!

**Comment by rubensworks** [comment]
Hmm, good catch @surilindur. There might be an edge-case that we're forgetting about.
Do you see any spec/integration tests failing by any chance?

We should probably delay merging this until we know for certain that this doesn't break anything in Comunica.

**Comment by surilindur** [comment]
This one seems to be happening consistently:
```
 FAIL  engines/query-sparql/test/QuerySparql-test.ts (51.179 s)
  ● System test: QuerySparql › foaf ontology broken link (no browser) › returns results with link recovery on [using full key]

    thrown: "Exceeded timeout of 20000 ms for a test.
    Add a timeout value to this test to increase the timeout, if this is a long-running test. See https://jestjs.io/docs/api#testname-fn-timeout."

      782 |   // We skip these tests in browsers due to CORS issues
      783 |   describe('foaf ontology broken link (no browser)', () => {
    > 784 |     it('returns results with link recovery on [using full key]', async() => {
          |     ^
      785 |       const result = <QueryBindings> await engine.query(`SELECT * WHERE {
      786 |     <http://xmlns.com/foaf/0.1/> a <http://www.w3.org/2002/07/owl#Ontology>.
      787 |   }`, {

      at engines/query-sparql/test/QuerySparql-test.ts:784:5
      at engines/query-sparql/test/QuerySparql-test.ts:783:3
      at Object.<anonymous> (engines/query-sparql/test/QuerySparql-test.ts:21:1)
```
But it looks like just a timeout at a first glance?

**Comment by surilindur** [comment]
To clarify, the following cases from the `MediatedQuadSource` seem to fail:

* MediatedQuadSource › A MediatedQuadSource instance with aggregated store › match › should match three chained sources when queried multiple times
* MediatedQuadSource › A MediatedQuadSource instance with aggregated store › match › should match three chained sources when queried multiple times in parallel
* MediatedQuadSource › A MediatedQuadSource instance with aggregated store › match › should match three chained sources when queried multiple times for multiple queries

And then that random timeout from `QuerySparql` tests.

**Comment by rubensworks** [comment]
> engines/query-sparql/test/QuerySparql-test.ts

@surilindur That one also fails in the CI sometimes. Honestly, I don't know what it is about. I've tried reproducing locally, but to no avail.

So no errors when running the integration and spec tests in query-sparql?

**Comment by surilindur** [comment]
The integration and spec tests in query-sparql seem to pass, yes.

**Comment by surilindur** [comment]
After trying out a lot of things, I realised the aggregate store option is toggled off in base Comunica by default. :grimacing:

Enabling the `aggregateStore` option with the `@default` JSDoc value [here](https://github.com/comunica/comunica/blob/20df744ee85764a8be3243a13686bb410f8ccb13/packages/actor-rdf-resolve-quad-pattern-hypermedia/lib/ActorRdfResolveQuadPatternHypermedia.ts#L130) caused the integration tests to fail.

<details><summary>Example failure in the spoiler here.</summary>

```
✖ SELECT - OPTIONAL FILTER EXISTS
  FILTER EXISTS works within an OPTIONAL block
  Error: Invalid query evaluation

        Query: SELECT * WHERE {
  ?s ?p ?o.
  OPTIONAL {
    FILTER EXISTS {
      ?s2 ?p ?o.
      FILTER (?s != ?s2)
    }
  }
}

        Data: [{"value":"https://comunica.github.io/manifest-ldf-tests/sparql/optional-filter-exists/2triples.ttl","type":"file"}]

        Result Source: https://comunica.github.io/manifest-ldf-tests/sparql/optional-filter-exists/result.srj

        Expected:
 [QueryResultBindings:
    Variables: [
  "?o",
  "?p",
  "?s"
]
    Bindings:  [
  {
    "?s": {
      "termType": "BlankNode",
      "value": "bc_0_b0_blank"
    },
    "?p": {
      "termType": "NamedNode",
      "value": "http://predicate"
    },
    "?o": {
      "termType": "NamedNode",
      "value": "http://object"
    }
  },
  {
    "?s": {
      "termType": "NamedNode",
      "value": "http://subject"
    },
    "?p": {
      "termType": "NamedNode",
      "value": "http://predicate"
    },
    "?o": {
      "termType": "NamedNode",
      "value": "http://object"
    }
  }
]
]

        Got:
 [QueryResultBindings:
    Variables: [
  "?o",
  "?p",
  "?s"
]
    Bindings:  [
  {
    "?s": {
      "termType": "BlankNode",
      "value": "bc_0_b35_blank",
      "skolemized": {
        "termType": "NamedNode",
        "value": "urn:comunica_skolem:source_0:b35_blank"
      }
    },
    "?p": {
      "termType": "NamedNode",
      "value": "http://predicate"
    },
    "?o": {
      "termType": "NamedNode",
      "value": "http://object"
    }
  }
]
]
```

</details>

Hopefully that helps. Sorry about the confusion earlier.

**Comment by rubensworks** [comment]
> After enabling the aggregateStore option with the @default JSDoc value [here](https://github.com/comunica/comunica/blob/20df744ee85764a8be3243a13686bb410f8ccb13/packages/actor-rdf-resolve-quad-pattern-hypermedia/lib/ActorRdfResolveQuadPatternHypermedia.ts#L130), most of the integration tests started failing.

That is to be expected. The aggregateStore is not compatible with TPF-like sources.

**Comment by surilindur** [comment]
Oh okay, thank you! I had no idea. Sorry. :frowning: What if I use query-sparql to query a single turtle document? Should the aggregated store support that?

With the aggregate store enabled, query-sparql fails to produce results for this, but without the aggregate store it works:
```
node bin/query.js https://comunica.github.io/manifest-ldf-tests/sparql/starwars-valuesbind/starwars.ttl -q 'SELECT ?person ?hair_colour WHERE {
  ?person <http://swapi.dev/documentation#name> "Luke Skywalker" ;
    <http://swapi.dev/documentation#hair_color> ?hair_colour .
}'
```
The same behaviour seems to happen with the link traversal repository, as well. With just one triple pattern it works, but with two or more it fails.

File: lib/StreamingStore.ts [hunk 6]
```diff
@@ -40,14 +40,65 @@ implements RDF.Source<Q>, RDF.Sink<RDF.Stream<Q>, EventEmitter> {
...
+        });
+
+        quad = stream.read();
       }
+      // If the stream has ended, all quads will be read from the quad stream, so we can end the storeImportStream.
+      if (streamEnded) {
+        storeImportStream.end();
+        return;
+      }
+      // If the stream hasn't ended, we recursively call this function to wait for the stream to become readable again.
```
Comment (rubensworks) [inline lib/StreamingStore.ts:96]: After reading this comment, I'm wondering if this recursion is still needed if we were to use `on` instead of `once` on line 51?

Related to this, we may also want to add a unit test to check for the case where the readable-event is invoked multiple times. The test on line 482 goes a bit in this direction, but I'm not sure if readable will actually be invoked multiple times there. We may have to resort to timeouts (which are simulated through jest, so we can manually step through them).

**Comment by rubensworks** [comment]
> The same behaviour seems to happen with the link traversal repository, as well. With just one triple pattern it works, but with two or more it fails.

@surilindur So this fails only with the new aggregate store from this PR, but not with the old aggregate store? Or does it fail on both?

**Comment by surilindur** [comment]
After checking it again, using the query from the previous comment here, with the aggregate store enabled:

* `query-sparql` fails to produce the expected result with *both* `master` and `fix-set-semantics-bug` branch, which probably means that is supposed to happen and everything is fine?
* `query-sparql-solid` does the same as `query-sparql`
* `query-sparql-link-traversal` produces the expected result with `master`, but fails to produce any results with `fix-set-semantics-bug`

So if the non-link-traversal engines are not supposed to work with the aggregate store for querying a turtle document, then everything is fine. Why the link traversal one works with master branch and not the fix, I do not know.

This is really confusing to me, so apologies for the spam and confusion.

**Comment by rubensworks** [comment]
> query-sparql fails to produce the expected result with both master and fix-set-semantics-bug branch, which probably means that is supposed to happen and everything is fine?

The aggregate store was designed mainly for link traversal, and is untested for regular querying. I would expect it to only break for TPF queries, but I may be forgetting about something, and the general breakage is not surprising.
So indeed, all is fine here.

>    query-sparql-solid does the same as query-sparql

👌
 
> query-sparql-link-traversal produces the expected result with master, but fails to produce any results with fix-set-semantics-bug

This is not ok indeed. And what exactly fails here? So the unit tests pass?

**Comment by surilindur** [comment]
I think I forgot to rebuild the link traversal engine because it seems to work now with the fix branch, as well.

But it does not make sense, because I definitely did build it with the fix branch when some of the queries failed.

File: lib/StreamingStore.ts (hunk 6 above)
Comment (rubensworks) [reply lib/StreamingStore.ts:96]: @maartyman Could you have a look to see if this is needed?
Would be good to get this one merged :-)

Comment (maartyman) [reply lib/StreamingStore.ts:96]: If `on` is used, test `handles duplicates in import (set-semantics)` fails. This is due to the next match being called on the store before the previous Quad is added. I can fix it by awaiting a timeout of 0 ms after adding the quad on line 68.

Comment (maartyman) [reply lib/StreamingStore.ts:96]: ^ nvm this is unreliable, sometimes the timing works out, sometimes it doesn't.

Comment (maartyman) [reply lib/StreamingStore.ts:96]: > Related to this, we may also want to add a unit test to check for the case where the readable-event is invoked multiple times. The test on line 482 goes a bit in this direction, but I'm not sure if readable will actually be invoked multiple times there. We may have to resort to timeouts (which are simulated through jest, so we can manually step through them).

It does invoke readable multiple times.

Comment (rubensworks) [reply lib/StreamingStore.ts:96]: Ok, thanks for checking, so no changes are needed then I guess?

Comment (maartyman) [reply lib/StreamingStore.ts:96]: No, everything should work!

Comment (rubensworks) [reply lib/StreamingStore.ts:96]: Ok, will merge then, thanks!
